src/service/viewport/viewport-impl.js

```diff
@@ -999,25 +1013,82 @@ export class Viewport {
     }
   }
 
+  /**
+   * Specifies a callback for measuring an element's fixed offset when the
+   * viewport's padding top changes.
+   *
+   * `measure` runs in a measure phase and takes:
+   *  - afterAnimation
+   *    a promise that resolves after a running translation is finished.
+   *  - lastPaddingTop in px.
+   *  - paddingTop in px.
+   *
+   * And returns a vertical offset at start of animation. End will always be 0.
+   *
+   * This tears down the element from the fixed layer (but does NOT remove from
+   * the transfer layer), so its offset needs to be managed independently
+   * afterwards.
+   *
+   * @param {!Element} element
+   * @param {!FixedElementMeasureFnDef} measure
+   */
+  setFixedElementMeasurer(element, measure) {
+    this.removeFromFixedLayer(
+        element, /* onlyTearDown */ true, /* keepOffset */ true);
+    this.fixedMeasurers_.push({element, measure});
+  }
+
   /**
    * @param {number} duration
    * @param {string} curve
    * @param {boolean} transient
    * @return {!Promise}
    * @private
    */
-  animateFixedElements_(duration, curve, transient) {
-    this.fixedLayer_.updatePaddingTop(this.paddingTop_, transient);
-    if (duration <= 0) {
-      return Promise.resolve();
-    }
-    // Add transit effect on position fixed element
-    const tr = numeric(this.lastPaddingTop_ - this.paddingTop_, 0);
-    return Animation.animate(this.ampdoc.getRootNode(), time => {
-      const p = tr(time);
-      this.fixedLayer_.transformMutate(`translateY(${p}px)`);
-    }, duration, curve).thenAlways(() => {
-      this.fixedLayer_.transformMutate(null);
+  updateFixedElementsOffset_(duration, curve, transient) {
+    const {
+      paddingTop_: paddingTop,
+      lastPaddingTop_: lastPaddingTop,
+    } = this;
+
+    this.fixedLayer_.updatePaddingTop(paddingTop, transient);
+
+    let doneDeferred;
+
+    return this.vsync_.measurePromise(() => {
+      return this.fixedMeasurers_.map(def => {
+        doneDeferred = doneDeferred || new Deferred();
```

Comment:
Why do any work if there were no `fixedMeasurers_` to begin with?

Reply:
it's essentially a no-op in that case, what is the issue?

Reply:
This entire `measurePromise` chain is a **very** expensive noop.

On this particular line, we already know there's going to need a deferred. Because if we didn't, there'd be no point in running any of this code? So this only serves to make the code less clear.

Reply:
Well, the deferred has to be created IFF there are `fixedMeasurers` and be resolved always, _regardless_ of whether there's an animation or not. But I guess we can decouple for performance. This was a deliberate choice for readability, not the other way around.